_posts/2025-04-19-navigating-pagination-with-hybrid-query.md

```diff
@@ -0,0 +1,229 @@
+The efficiency of pagination depends on how evenly your data is distributed across shards. You can check this distribution using the `_cat/shards` API. If the documents are equally distributed amongst shards, then even a smaller `pagination_depth` can provide more hybridized search results per subquery.
+
+For example:
+* With `pagination_depth = 20` across three shards, you can hybridize up to 60 results per subquery.
+* This assumes each shard contains at least 20 results per subquery (thus, the data is evenly distributed).
+* If there are no duplicate entries between subqueries, you can paginate through up to 120 hybridized search results.
+
+### Handle uneven data distribution
+
+Custom routing can lead to uneven data distribution across shards, resulting in:
+* Some shards returning zero results, reducing the total number of hybridized results.
+
+### Some shards may contain all the relevant results
+
+If a shard contains more than 20 results per subquery, it can still return only 20 results (based on `pagination_depth`).
+
+As a result, the coordinator node will receive fewer results for hybridization, reducing the final search result size.
+
+### Recommendations
+
+Follow these best practices to optimize hybrid search for efficiency, ensuring high relevance while maintaining performance:
+
+* **Avoid deep pagination**: Hybrid search is optimized for relevance, not exhaustive pagination. For deep pagination needs, consider using traditional search methods like `bool` or `match` queries.
+* **Distribute data evenly**: Balance shards during indexing for better query efficiency and to minimize the need for high `pagination_depth` values.
+
+## Limitations
+
+You may see this error while navigating between pages: "Reached end of search results. Increase pagination_depth value to see more results."
+
+As noted in [Key takeaways](#key-takeaways), maintaining a consistent `pagination_depth` ensures stable results but limits you to a fixed result set. When you reach the last available page for your specified `pagination_depth`, OpenSearch will return an error indicating that the end of the search results has been reached. At this point, you have exhausted the search results for the given `pagination_depth`. To retrieve additional results beyond this limit, you must increase the `pagination_depth`, effectively expanding the search reference by including more results in the final result set.
+
+## Wrapping up
+
+In this blog post, we've explored how `pagination_depth` resolves the ground truth issues that occur when using `from` and `size` parameters alone for pagination. We also examined how `pagination_depth` impacts the performance of retrieving the top pages of the final search result. As discussed, applying pagination to retrieve a higher number of pages can degrade search performance, as it increases both computation time and cost. To help you optimize performance, we've provided best practices for using `pagination_depth` with hybrid search. 
```

Comment:
```suggestion
In this blog post, we explored how `pagination_depth` resolves the ground truth issues that occur when using only the `from` and `size` parameters for pagination. We also examined how `pagination_depth` affects performance when retrieving the top pages of the final search result. As discussed, applying pagination to retrieve a higher number of pages can degrade search performance because it increases both computation time and costs. To help you optimize performance, we've provided best practices for using `pagination_depth` with hybrid search. 
```